### ISSUE TITLE
index.py errors

### ISSUE THREAD

**Issue by Ivypillar** [kjenike/Panagram#5]
Hi, this looks like a really interesting program, k-mers and pangenomes together at last. The latest fix solved some installation issues i had, but I still encountered problems.

The pip installation doesn't run unless I install some dependencies first with conda. e.g. conda install kmc tabix pysam pip mash

Initially I had an error in index.py line 271, caused by the way I'd named the assemblies in the config file. I had haplotypes labelled in the format: sample1-assembly.1, sample1-assembly.2 in the style of cactus config files. Altering these to sample1-assembly_1, sample1-assembly_2, fixed the error.

Following this I had errors in index.py due to paths for kmc and mash. I could resolve this error by removing all instances of "{EXTRA_DIR}/" in the index.py, forcing the use of kmc and mash commands available in the path. Or alternatively, including symbolic links to the conda versions of kmc & kmc_tools in the "site-packages/panagram/extra/" folder.

Following this I had errors in index.py due to kmc_tools', 'complex', './kmc_bitvec/0.opdef.txt' returned non-zero exit status 1. Tracing the errors back suggested that assembly names in the config file was the underlying cause again, this time because some assemblies included the hyphen character, I switched these to underscores as with the previous bug. e.g. sample1_assembly_1, sample1_assembly_2.

Following this I had another error linked to the "{EXTRA_DIR}/" path, this time loading "py_kmc_api". It turns out the conda version of KMC doesn't install this. I had issues install KMC and py_kmc_api from GitHub so I'm a bit stuck for now.

**Comment by skovaka** [comment]
Thank you for the feedback! Most dependencies should be installed via pip (see pyproject.toml), except for samtools and tabix which I've added to the README. I've also added a minimum python version requirement and a description of the genome ID limitations. I should add an option to use the system default kmc and mash binaries, but py_kmc_api must be compiled for the specific OS and python version. 

Could you share the output of `pip install -v .`? The output can be quite long, so you may want to run `pip install -v . > pip_out.txt 2>&1` to capture all output into pip_out.txt.

It might be an issue with your setuptools version, in which case you could fix it with `pip install --upgrade setuptools`, but I would like to support older versions so sharing your output would be very helpful.

**Comment by Ivypillar** [comment]
I couldn't get pip to install KMC correctly, however I was able to get panagram working by cloning the KMC GitHub repo and building the python api with make, I then copied the whole KMC folder into the panagram repo KMC folder. I then re-ran the "pip install -v ." command. This produced a new error linked to the KMC API which seemed to be returning an object rather than a list of values. I made a bit of a guess as to which attribute was needed for the np.array line and modified the index.py line from:

pac = np.array(vec, dtype="uint32")

to 

pac = np.array(vec.value, dtype="uint32")

The indexing and viewing now seems to work for the example data. I'm still having some problems with my larger genome dataset though. I will post here if I fix it.

I've attached the pip logs from before and after the crude fix.

[panagram_pip_fixed.log](https://github.com/kjenike/panagram/files/10677094/panagram_pip_fixed.log)
[panagram_pip.log](https://github.com/kjenike/panagram/files/10677095/panagram_pip.log)

**Comment by skovaka** [comment]
Ah, did you clone the repository with the `--recursive` option? We actually rely on a fork of KMC included as a git submodule, which has an optimization in binding the "vec" to a numpy array. You can manually download the submodule via `git submodule update --init`.

For your large dataset, does it contain more than 32 genomes? Panagram is currently more reliable for <= 32 genomes, and we are working on identifying problems for larger datasets. I'll add a note about that in the README

**Comment by Ivypillar** [comment]
Ah sorry this is my mistake, I completely missed the --recursive option on the readme page! I just did a regular git clone out of habit, rather than copying the commands from the readme. This fixed the issue I was having with KMC.

I had another error with panagram view. I made sure that I reinstalled with --recursive, but it still persists. The traceback is to view.py line 686, and essentially it looks like "rep_colors" runs out of colours. I can surpress the error by adding a bunch of extra "grey" entries to the original list declaration on line 647-8, but I'm not sure how this will affect the output. This happened when running with a test set of three moderate sized eukaryotic genomes.

The full dataset is 36 genomes, roughly 400Mbp each.
